pallets/pools/src/tranche.rs

```diff
@@ -0,0 +1,386 @@
+
+	pub fn acc_investments(&self) -> Result<Balance, DispatchError> {
+		self.0
+			.iter()
+			.fold(Some(Balance::zero()), |sum, tranche| {
+				sum.and_then(|acc| acc.checked_add(&tranche.outstanding_invest_orders))
+			})
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn redemptions(&self) -> Vec<Balance> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.outstanding_redeem_orders)
+			.collect()
+	}
+
+	pub fn acc_redemptions(&self) -> Result<Balance, DispatchError> {
+		self.0
+			.iter()
+			.fold(Some(Balance::zero()), |sum, tranche| {
+				sum.and_then(|acc| acc.checked_add(&tranche.outstanding_redeem_orders))
+			})
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn calculate_weights(&self) -> Vec<(Weight, Weight)> {
+		let n_tranches: u32 = self.0.len().try_into().expect("MaxTranches is u32");
+		let redeem_starts = 10u128.checked_pow(n_tranches).unwrap_or(u128::MAX);
+
+		self.0
+			.iter()
+			.map(|tranche| {
+				(
+					10u128
+						.checked_pow(
+							n_tranches
+								.checked_sub(tranche.seniority)
+								.unwrap_or(u32::MAX),
+						)
+						.unwrap_or(u128::MAX)
+						.into(),
+					redeem_starts
+						.checked_mul(10u128.pow(tranche.seniority.saturating_add(1)).into())
+						.unwrap_or(u128::MAX)
+						.into(),
+				)
+			})
+			.collect()
+	}
+
+	pub fn risk_buffers(&self) -> Vec<Perquintill> {
+		self.0
+			.iter()
+			.map(|tranche| tranche.min_risk_buffer)
+			.collect()
+	}
+}
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug, Default, TypeInfo)]
+pub struct EpochExecutionTranches<Balance, BalanceRatio, Weight>(
+	pub(super) Vec<EpochExecutionTranche<Balance, BalanceRatio, Weight>>,
+);
+
+impl<Balance, BalanceRatio, Weight> EpochExecutionTranches<Balance, BalanceRatio, Weight>
+where
+	Balance: Zero + Copy + BaseArithmetic + Unsigned + From<u64>,
+	Weight: Copy + From<u128>,
+	BalanceRatio: Copy,
+{
+	pub fn num_tranches(&self) -> usize {
+		self.0.len()
+	}
+
+	pub fn into_tranches(self) -> Vec<EpochExecutionTranche<Balance, BalanceRatio, Weight>> {
+		self.0
+	}
+
+	pub fn as_tranche_slice(&self) -> &[EpochExecutionTranche<Balance, BalanceRatio, Weight>] {
+		self.0.as_slice()
+	}
+
+	pub fn as_mut_tranche_slice(
+		&mut self,
+	) -> &mut [EpochExecutionTranche<Balance, BalanceRatio, Weight>] {
+		self.0.as_mut_slice()
+	}
+
+	pub fn prices(&self) -> Vec<BalanceRatio> {
+		self.0.iter().map(|tranche| tranche.price).collect()
+	}
+
+	pub fn supplies_with_fulfillment(
+		&self,
+		fulfillments: &[TrancheSolution],
+	) -> Result<Vec<Balance>, DispatchError> {
+		self.0
+			.iter()
+			.zip(fulfillments)
+			.map(|(tranche, solution)| {
+				tranche
+					.supply
+					.checked_add(&solution.invest_fulfillment.mul_floor(tranche.invest))
+					.and_then(|value| {
+						value.checked_sub(&solution.redeem_fulfillment.mul_floor(tranche.redeem))
+					})
+			})
+			.collect::<Option<Vec<_>>>()
+			.ok_or(ArithmeticError::Overflow.into())
+	}
+
+	pub fn acc_supply_with_fulfillment(
```

Comment:
Can we also rename the `acc_` methods here? I realize you've somewhat copied this from the previous implementation. But I don't think `acc_` is clear, I wouldn't immediately know what it means.